src/ledger/LedgerManagerImpl.cpp

```diff
@@ -1573,6 +1576,430 @@ LedgerManagerImpl::prefetchTransactionData(AbstractLedgerTxnParent& ltx,
     }
 }
 
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+
+ThreadEntryMap
+LedgerManagerImpl::collectEntries(AbstractLedgerTxn& ltx, Thread const& txs)
+{
+    ThreadEntryMap entryMap;
+    auto getEntries = [&](TransactionFrameBasePtr tx,
+                          MutableTxResultPtr resPayload,
+                          xdr::xvector<LedgerKey> const& keys,
+                          uint32_t& readBytes) {
+        for (auto const& lk : keys)
+        {
+            if (!resPayload->isSuccess())
+            {
+                // This tx has already failed, so no need to load anything
+                break;
+            }
+
+            auto ltxe = ltx.loadWithoutRecord(lk);
+            if (ltxe)
+            {
+                auto entrySize =
+                    static_cast<uint32_t>(xdr::xdr_size(ltxe.current()));
+                readBytes = add_sat(readBytes, entrySize);
+
+                entryMap.emplace(lk, ThreadEntry{ltxe.current(), false});
+
+                if (isSorobanEntry(lk))
+                {
+                    auto ttlKey = getTTLKey(lk);
+                    auto ttlLtxe = ltx.loadWithoutRecord(ttlKey);
+                    // TTL entry must exist
+                    releaseAssertOrThrow(ttlLtxe);
+
+                    entryMap.emplace(ttlKey,
+                                     ThreadEntry{ttlLtxe.current(), false});
+                }
+            }
+            else
+            {
+                entryMap.emplace(lk, ThreadEntry{std::nullopt, false});
+
+                if (isSorobanEntry(lk))
+                {
+                    auto ttlKey = getTTLKey(lk);
+                    auto ttlLtxe = ltx.loadWithoutRecord(ttlKey);
+                    // TTL entry must not exist
+                    releaseAssertOrThrow(!ttlLtxe);
+                    entryMap.emplace(ttlKey, ThreadEntry{std::nullopt, false});
+                }
+            }
+
+            if (tx->sorobanResources().readBytes < readBytes)
+            {
+                // fast fail this tx now that we know the readBytes limit has
+                // been exceeded.
+                resPayload->setInnermostResultCode(txFAILED);
+
+                auto& sorobanOpResult = resPayload->getOpResultAt(0);
+                auto opType = tx->getRawOperations().at(0).body.type();
+                switch (opType)
+                {
+                case INVOKE_HOST_FUNCTION:
+                    sorobanOpResult.tr().invokeHostFunctionResult().code(
```

Comment:
We also need to populate the respective events here, don't we?

Reply:
Updated